lib/src/test/java/io/ably/lib/test/realtime/RealtimeJWTTest.java

```diff
@@ -0,0 +1,285 @@
+			connectionWaiter.waitFor(ConnectionState.connected);
+			assertEquals("Connected state was NOT reached", ConnectionState.connected, ablyRealtime.connection.state);
+
+			/* attach to channel and verify attached state */
+			Channel channel = ablyRealtime.channels.get(channelName);
+			channel.attach();
+			new ChannelWaiter(channel).waitFor(ChannelState.attached);
+
+			/* publish, verify that it succeeds then close */
+			final Message message = new Message(messageName, null);
+			channel.publish(message, new CompletionListener() {
+				@Override
+				public void onSuccess() {
+					System.out.println("Message " + messageName + " published successfully");
+					ablyRealtime.close();
+				}
+
+				@Override
+				public void onError(ErrorInfo reason) {
+					fail("Publish should not fail");
+				}
+			});
+			connectionWaiter.waitFor(ConnectionState.closed);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail();
+		}
+	}
+
+	/**
+	 * Request a JWT with a ttl of 5 seconds and
+	 * verify the correct error and message in the disconnected state change.
+	 */
+	@Test
+	public void auth_jwt_with_token_that_expires() {
+		try {
+			/* create ably realtime with JWT token that expires in 5 seconds */
+			realtimeOptions.token = getToken(mergeParams(keys, shortTokenTtl), null);
+			assertNotNull("Expected token value", realtimeOptions.token);
+			final AblyRealtime ablyRealtime = new AblyRealtime(realtimeOptions);
+
+			/* wait for connected state */
+			ConnectionWaiter connectionWaiter = new ConnectionWaiter(ablyRealtime.connection);
+			connectionWaiter.waitFor(ConnectionState.connected);
+			assertEquals("Connected state was NOT reached", ConnectionState.connected, ablyRealtime.connection.state);
+
+			/* Verify the expected error reason when disconnected */
+			ablyRealtime.connection.once(ConnectionEvent.disconnected, new ConnectionStateListener() {
+
+				@Override
+				public void onConnectionStateChanged(ConnectionStateChange stateChange) {
+					assertEquals("Unexpected connection stage change", 40142, stateChange.reason.code);
+					assertEquals("Unexpected error message", "Key/token status changed (expire)", stateChange.reason.message);
+					ablyRealtime.close();
+				}
+			});
+			connectionWaiter.waitFor(ConnectionState.disconnected);
+			connectionWaiter.waitFor(ConnectionState.closed);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail();
+		}
+	}
+
+	/**
+	 * Request a JWT with a ttl of 35 seconds and
+	 * verify that the client reauths without going through a disconnected state. (RTC8a4)
+	 */
+	@Test
+	public void auth_jwt_with_client_than_reauths_without_disconnecting() {
+		try {
+			final String[] tokens = new String[1];
+			final boolean[] authMessages = new boolean[] { false };
+
+			/* create ably realtime with authUrl and params that include a ttl of 35 seconds */
+			DebugOptions options = new DebugOptions(testVars.keys[0].keyStr);
+			options.environment = "sandbox";
```

Comment:
Please don't specify the environment explicitly. That stops the tests from being run against any other environment. Use `createOptions()` or something to get options where the `environment` is set according to the test setup.